lib/Pingmachine/Probe/HTTPing.pm

```diff
@@ -0,0 +1,256 @@
+        }
+    );
+}
+
+sub _kill_current_job {
+    my ($self) = @_;
+
+    # Kill httping, if still running
+    my $job = $self->current_job;
+    for my $url (keys %{$job->{url2order}}) {
+        if($job->{pid}{$url}) {
+            # Check that we are killing the process we started and not an innocent bystander
+            my $cmd_match = 0;
+            if (open(proc_fh, "/proc/$job->{pid}{$url}/cmdline")) {
+                $cmd_match = (join('', readline(proc_fh)) eq $job->{cmd}{$url});
+                close(proc_fh);
+            }
+            if($cmd_match && kill(0, $job->{pid}{$url})) {
+                $log->warning("killing unfinished httping process (step: ".$self->step.", pings: ".$self->pings.", offset: ".$self->time_offset().")");
+                kill 9, $job->{pid}{$url};
+                $job->{pid}{$url} = undef;
+	    }
+            elsif($job->{output}{$url}) {
+                $log->warning("httping has finished, but we didn't notice... collecting (step: ".$self->step.", pings: ".$self->pings.", offset: ".$self->time_offset.")");
+                $self->_collect_current_job();
+            }
+            else {
+                $log->warning("httping has finished, but we didn't notice... no output found (?)");
+            }
+        }
+    }
+}
+
+sub _collect_current_job {
+    my ($self) = @_;
+    my $text;
+
+    my $job = $self->current_job;
+    $self->current_job({});
+    my %results;
+
+    # Parse httping report
+    for my $url (keys %{$job->{url2order}}) {
+        # Do nothing, if httping didn't run yet or if job has been already collected
+        return unless $job->{output}{$url};
+        
+        my $raw_text = $job->{output}{$url};
+        # sample output
+        # PING neverssl.com:80 (/):
+        #	connected to neverssl.com:80 (524 bytes), seq=0 time= 10.86 ms
+        #	connected to neverssl.com:80 (524 bytes), seq=1 time= 15.46 ms
+        #	--- http://neverssl.com/ ping statistics ---
+        #	2 connects, 2 ok, 0.00% failed, time 2027ms
+        #	round-trip min/avg/max = 10.9/13.2/15.5 ms
+        # parse line by line httping output and format it as with fping
+        my @lines = split /\n/, $raw_text;
+        my $values;
+        foreach my $line (@lines) {
+            # if the line contains an error (i.e. timeout, connection refused, ...) add a -
+            if ($line =~ /could not connect|timeout|short read/) {
+                $values .= '- ';
+            }
+            # if the line contains the httping result add the number
+            if ($line =~ /time=\s*(\d\d*.\d\d*) ms/) {
+                $values .= "$1 ";
+            }
+            # ignore all the rest
+        }
+        $values =~ s/\s+$//;
+        $text .= "$url : $values\n";
+
+    }
+
+    while($text !~ /\G\z/gc) {
+        if($text =~ /\G(\S+)[ \t]+:/gc) {
+            my $url = $1;
+            my @data;
+            while($text =~ /\G[ \t]+([-\d\.]+)/gc) {
```

Comment:
I dislike if the data is encoded and parsed again in the same code. Better keep it structured all the time.